src/query/functions/linear/histogram_quantile.go

```diff
@@ -0,0 +1,376 @@
+		}
+
+		sort.Sort(buckets)
+		maxBound := buckets.buckets[len(buckets.buckets)-1].upperBound
+		if !math.IsInf(maxBound, 1) {
+			delete(bucketMap, k)
+		}
+	}
+}
+
+func bucketQuantile(q float64, buckets []bucketValue) float64 {
+	// NB: some valid buckets may have been purged if the values at the current
+	// step for that series are not present.
+	if len(buckets) < 2 {
+		return math.NaN()
+	}
+
+	// NB: similar situation here if the max bound bucket does not have a value
+	// at this point, it is necessary to re-check.
+	if !math.IsInf(buckets[len(buckets)-1].upperBound, 1) {
+		return math.NaN()
+	}
+
+	rank := q * buckets[len(buckets)-1].value
+
+	bucketIndex := sort.Search(len(buckets)-1, func(i int) bool {
+		return buckets[i].value >= rank
+	})
+
+	if bucketIndex == len(buckets)-1 {
+		return buckets[len(buckets)-2].upperBound
+	}
+
+	if bucketIndex == 0 && buckets[0].upperBound <= 0 {
+		return buckets[0].upperBound
+	}
+
+	var (
+		bucketStart float64
+		bucketEnd   = buckets[bucketIndex].upperBound
+		count       = buckets[bucketIndex].value
+	)
+
+	if bucketIndex > 0 {
+		bucketStart = buckets[bucketIndex-1].upperBound
+		count -= buckets[bucketIndex-1].value
+		rank -= buckets[bucketIndex-1].value
+	}
+
+	return bucketStart + (bucketEnd-bucketStart)*rank/count
+}
+
+// Process the block
+func (n *histogramQuantileNode) Process(ID parser.NodeID, b block.Block) error {
+	stepIter, err := b.StepIter()
+	if err != nil {
+		return err
+	}
+
+	meta := stepIter.Meta()
+	seriesMetas := utils.FlattenMetadata(meta, stepIter.SeriesMeta())
+	bucketedSeries := gatherSeriesToBuckets(seriesMetas)
+
+	q := n.op.q
+	if q < 0 || q > 1 {
+		return processInvalidQuantile(q, bucketedSeries, meta, stepIter, n.controller)
+	}
+
+	return processValidQuantile(q, bucketedSeries, meta, stepIter, n.controller)
+}
+
+func setupBuilder(
+	bucketedSeries bucketedSeries,
+	meta block.Metadata,
+	stepIter block.StepIter,
+	controller *transform.Controller,
+) (block.Builder, error) {
+	metas := make([]block.SeriesMeta, len(bucketedSeries))
+	idx := 0
+	for _, v := range bucketedSeries {
+		metas[idx] = block.SeriesMeta{
+			Tags: v.tags,
+		}
+
+		idx++
+	}
+
+	meta.Tags, metas = utils.DedupeMetadata(metas)
+	builder, err := controller.BlockBuilder(meta, metas)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = builder.AddCols(stepIter.StepCount()); err != nil {
+		return nil, err
+	}
+
+	return builder, nil
+}
+
+func processValidQuantile(
+	q float64,
+	bucketedSeries bucketedSeries,
+	meta block.Metadata,
+	stepIter block.StepIter,
+	controller *transform.Controller,
+) error {
+	sanitizeBuckets(bucketedSeries)
+
+	builder, err := setupBuilder(bucketedSeries, meta, stepIter, controller)
+	if err != nil {
+		return err
+	}
+
+	for index := 0; stepIter.Next(); index++ {
+		step := stepIter.Current()
+		values := step.Values()
+		bucketValues := make([]bucketValue, 0, initIndexBucketLength)
+
+		aggregatedValues := make([]float64, 0, len(bucketedSeries))
```

Comment:
You should be able to just pre-allocate this with a length and index in below when adding to it. Looks like you always add to it.

Reply:
Will do; was fine using `append` here since didn't want to add a tracking `idx` variable